File: java/VendingMachine/src/com/ravendarque/Inventory.java
package com.ravendarque;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

class Inventory {

    private final ArrayList<Item> items = new ArrayList<>();

    List<Item> getItems() {

        return items;
    }

    List<Item> getItemsByType(Class clazz) {

        return items.stream()
                    .filter(item -> item.getClass() == clazz)
                    .collect(Collectors.toList());
    }

    void removeItem(Item item) {

        items.remove(item);
    }

    void addItem(Item testItem) {

        items.add(testItem);
    }
}

File: C++/VendingMachine/VendingMachine/Inventory.cpp
#include "pch.h"
#include "Inventory.h"
#include <vector>
#include <algorithm>

std::vector<Item*> Inventory::GetItems() const
{
	return Items;
}

void Inventory::AddItem(Item& item)
{
	Items.push_back(&item);
}

bool Inventory::operator==(const Inventory& rhs) const
{
	if (Items.size() != rhs.GetItems().size())
		return false;

	for (auto i = 0U;i < Items.size();i++)
		if (&Items.at(i) != &rhs.GetItems().at(i)) return false;

	return true;
}

void Inventory::RemoveItem(Item& item)
{
	Items.erase(std::remove(Items.begin(), Items.end(), &item), Items.end());
}
